src/liquidity_mining/invariants.rs

```diff
@@ -0,0 +1,163 @@
+use crate::{assert_eq_approx, types::Balance};
+use proptest::prelude::*;
+use sp_arithmetic::traits::One;
+use sp_arithmetic::{
+    traits::{CheckedAdd, CheckedDiv, CheckedMul, CheckedSub},
+    FixedPointNumber, FixedU128,
+};
+
+fn periods() -> impl Strategy<Value = u32> {
+    0..=u32::MAX
+}
+
+fn initial_reward_percentage() -> impl Strategy<Value = u128> {
+    1..=999_999_999_999_999_999_u128 //< FixedU128::one()
+}
+
+fn scale_coef() -> impl Strategy<Value = u32> {
+    0..=(u32::MAX / 2) //half of periods() so we are able to test `periods > scale_coef` case
+}
+
+proptest! {
+    #![proptest_config(ProptestConfig::with_cases(1_000))]
+    #[test]
+    fn calculate_loyalty_multiplier(
+        periods in periods(),
+        initial_reward_percentage in initial_reward_percentage(),
+        scale_coef in scale_coef(),
+    ) {
+        let b = FixedU128::from_inner(initial_reward_percentage);
+        let multiplier = crate::liquidity_mining::calculate_loyalty_multiplier(
+            periods,
+            b,
+            scale_coef,
+        ).unwrap();
+
+        assert!(multiplier < FixedU128::one(), "loyalty multiplier is >= 1");
+
+        //b + (1 - b)/2;
+        let bound = FixedU128::one()
+                .checked_sub(&b).unwrap()
+                .checked_div(&FixedU128::from(2_u128)).unwrap()
+                .checked_add(&b).unwrap();
+
+        if periods < scale_coef {
+            assert!(multiplier >= b && multiplier <= bound);
+        } else if periods > scale_coef {
+            assert!(multiplier >= bound && multiplier <= FixedU128::one());
+        } else {
+            assert_eq!(multiplier, bound);
+        }
+
+        //LoyaltyFactor * (t + tb + scaleCoef*(b + 1)) == t + tb + b*scaleCoef*(b + 1)
+        let t = FixedU128::from(TryInto::<u128>::try_into(periods).unwrap());
+        let t_add_tb =  b
+            .checked_mul(&t).unwrap()
+            .checked_add(&t).unwrap();
+
+        let scale_coef_mul_b_add_one = FixedU128::one()
+            .checked_add(&b).unwrap()
+            .checked_mul(&FixedU128::from(scale_coef as u128)).unwrap();
+
+        let lhs = t_add_tb
+            .checked_add(&scale_coef_mul_b_add_one).unwrap()
+            .checked_mul(&multiplier).unwrap();
+
+        let rhs = b
+            .checked_mul(&scale_coef_mul_b_add_one).unwrap()
+            .checked_add(&t_add_tb).unwrap();
+
+        let tolerance = FixedU128::from_inner(20_000_000_000); //0.000_000_02
+        assert_eq_approx!(
+            lhs,
+            rhs,
+            tolerance,
+            "loyalty multiplier, LoyaltyFactor * (t + tb + scaleCoef*(b + 1)) == t + tb + b*scaleCoef*(b + 1)"
+        );
+    }
+
+    #[test]
+    fn calculate_loyalty_multiplier_should_eqaul_to_bound_when_period_is_equal_scale_coef(
+        periods in periods(),
+        initial_reward_percentage in initial_reward_percentage(),
+    ) {
+        //This case never hit by test above, that's why there is special test only for this
```

Comment:
Definitely agree with splitting this out as a separate test, it validates the meaning of the scale_coef parameter